File: game/card.go
package game

type Suit string

const (
	Spades   Suit = "Spades"
	Hearts   Suit = "Hearts"
	Diamonds Suit = "Diamonds"
	Clubs    Suit = "Clubs"
)

const NUM_SUITS = 4

type Rank string

const (
	Seven Rank = "7"
	Eight Rank = "8"
	Nine  Rank = "9"
	Ten   Rank = "10"
	Jack  Rank = "J"
	Queen Rank = "Q"
	King  Rank = "K"
	Ace   Rank = "A"
)

const NUM_CARD_VALUES = 8

type Card struct {
	Suit Suit `json:"suit"`
	Rank Rank `json:"rank"`
}

func (c *Card) String() string {
	return string(c.Rank) + " of " + string(c.Suit)
}

var naturalOrderIndex = map[Rank]int{
	Seven: 0,
	Eight: 1,
	Nine:  2,
	Ten:   3,
	Jack:  4,
	Queen: 5,
	King:  6,
	Ace:   7,
}

var trumpTrickOrderIndex = map[Rank]int{
	Seven: 0,
	Eight: 1,
	Queen: 2,
	King:  3,
	Ten:   4,
	Ace:   5,
	Nine:  6,
	Jack:  7,
}

var nonTrumpTrickOrderIndex = map[Rank]int{
	Seven: 0,
	Eight: 1,
	Nine:  2,
	Jack:  3,
	Queen: 4,
	King:  5,
	Ten:   6,
	Ace:   7,
}

func (r *Rank) NaturalOrder() int {
	return naturalOrderIndex[*r]
}

func (r *Rank) TrickOrder(isTrump bool) int {
	if isTrump {
		return trumpTrickOrderIndex[*r]
	}
	return nonTrumpTrickOrderIndex[*r]
}

func (r *Rank) GetNonTrumpPoints() int {
	points := map[Rank]int{
		Seven: 0,
		Eight: 0,
		Nine:  0,
		Jack:  2,
		Queen: 3,
		King:  4,
		Ten:   10,
		Ace:   11,
	}
	return points[*r]
}

func (r *Rank) GetTrumpPoints() int {
	points := map[Rank]int{
		Seven: 0,
		Eight: 0,
		Queen: 3,
		King:  4,
		Ten:   10,
		Ace:   11,
		Nine:  14,
		Jack:  20,
	}
	return points[*r]
}

func Less(r1, r2 Rank, isTrump bool) bool {
	return r1.TrickOrder(isTrump) < r2.TrickOrder(isTrump)
}

File: client/app/client/card.ts
import { IsEnum } from "class-validator";

export enum Suit {
	Spades = "Spades",
	Hearts = "Hearts",
	Diamonds = "Diamonds",
	Clubs = "Clubs"
}

export enum Rank {
	Seven = "7",
	Eight = "8",
	Nine = "9",
	Ten = "10",
	Jack = "J",
	Queen = "Q",
	King = "K",
	Ace = "A"
}

const trumpOrder = [Rank.Jack, Rank.Nine, Rank.Ace, Rank.Ten, Rank.King, Rank.Queen, Rank.Eight, Rank.Seven];
const nonTrumpOrder = [Rank.Ace, Rank.Ten, Rank.King, Rank.Queen, Rank.Jack, Rank.Nine, Rank.Eight, Rank.Seven];

export class Card {
	@IsEnum(Suit)
	suit: Suit;

	@IsEnum(Rank)
	rank: Rank;

	constructor(suit: Suit, rank: Rank) {
		this.suit = suit;
		this.rank = rank;
	}

	compare(other: Card, trumpSuit: Suit): number {
		const thisOrder = this.suit === trumpSuit ? trumpOrder : nonTrumpOrder;
		const otherOrder = other.suit === trumpSuit ? trumpOrder : nonTrumpOrder;

		const thisRankIndex = thisOrder.indexOf(this.rank);
		const otherRankIndex = otherOrder.indexOf(other.rank);

		if (this.suit === other.suit) {
			return otherRankIndex - thisRankIndex;
		}

		if (this.suit === trumpSuit) {
			return 1;
		}

		if (other.suit === trumpSuit) {
			return -1;
		}

		return 0;
	}

	equals(other: Card) {
		return this.suit === other.suit && this.rank === other.rank;
	}
}
